### PR TITLE
Running upstream tests on Azure pipelines

### PR DESCRIPTION
**Pull request by seocam**
In order to run the tests on Azure it has been added the option to run our tests inside containers as well.

When using ansible with containers the inventory file won't have the fqdn but just the container name. 

Due to that the tests have been changed to use `{{ ansible_fqdn }}` instead of `{{ groups.ipaserver[0] }}` (which works fine since all tests modify use `hosts: ipaservers`).

### CODE REVIEW

**Comment by seocam** [comment]
There still 7 tests failing. I need some help to get it fixed.

**Comment by rjeffman** [comment]
Some, if not all test failures in the provided link are due to DNS failures. I'm still not sure if this is a bug in the modules or an environment configuration.

On my configuration, `/etc/resolv.conf`for the IPA servers point to itself (127.0.0.1 or the machine IP), and all tests pass. If the machine IP is not in the name servers list, I see the same errors is tests.

**Comment by seocam** [comment]
> Some, if not all test failures in the provided link are due to DNS failures. I'm still not sure if this is a bug in the modules or an environment configuration.
> 
> On my configuration, `/etc/resolv.conf`for the IPA servers point to itself (127.0.0.1 or the machine IP), and all tests pass. If the machine IP is not in the name servers list, I see the same errors is tests.

Thanks @rjeffman. I've added a temp commit (b4fbe5575317f7166bca2f18be9c71385cd8f7a5) setting resolv.conf to see how it goes.

**Comment by seocam** [comment]
We got all green: https://dev.azure.com/seocam/seocam-ansible-freeipa/_build/results?buildId=113&view=ms.vss-test-web.build-test-results-tab
Thanks @rjeffman!! :D

**Comment by t-woerner** [comment]
Please add this information about the changes and why they are needed to the commits. Currently the commits only have a subject.

**Comment by seocam** [comment]
> Please add information about the changes and why they are needed to the commits. Currently the commits only have a subject.

Done.

File: tests/ca-less/install_server_without_ca.yml [hunk 1]
```diff
@@ -1,15 +1,15 @@
...
+  gather_facts: true
 
   tasks:
   - name: Run generate-certificates.sh
     command: >
       /bin/bash
       generate-certificates.sh create
-      "{{ groups.ipaserver[0] }}"
-      "{{ ipaserver_domain | default(groups.ipaserver[0].split('.')[1:] | join ('.')) }}"
+      "{{ ansible_fqdn }}"
```
Comment (t-woerner) [inline tests/ca-less/install_server_without_ca.yml:11]: This is not correct, the task is executed on the controller. `ansible_fqdn` is not `groups.ipaserver[0]`.

Comment (seocam) [reply tests/ca-less/install_server_without_ca.yml:11]: Same case in tests/external-signed-ca-with-automatic-copy/install-server-with-external-ca-with-automatic-copy.yml.
I'll look for the playbooks running in localhost and undo the changes in those.

Comment (seocam) [reply tests/ca-less/install_server_without_ca.yml]: I've updated the PR undoing the changes for the localhost cases.

File: tests/external-signed-ca-with-automatic-copy/install-server-with-external-ca-with-automatic-copy.yml [hunk 2]
```diff
@@ -28,8 +27,8 @@
   hosts: ipaserver
   become: true
   vars:
-    ipaserver_external_cert_files_from_controller: "{{ groups.ipaserver[0] + '-chain.crt' }}"
-    #ipaserver_external_ca_file: "{{ groups.ipaserver[0] + '-cacert.asc' }}"
+    ipaserver_external_cert_files_from_controller: "{{ ansible_fqdn + '-chain.crt' }}"
```
Comment (t-woerner) [inline tests/external-signed-ca-with-automatic-copy/install-server-with-external-ca-with-automatic-copy.yml:30]: This is run on `localhost`. The use of `ansible_fqdn` is changing the names used for the certicate.

File: tests/external-signed-ca-with-automatic-copy/install-server-with-external-ca-with-automatic-copy.yml [hunk 3]
```diff
@@ -28,8 +27,8 @@
...
+    #ipaserver_external_ca_file: "{{ ansible_fqdn + '-cacert.asc' }}"
```
Comment (t-woerner) [inline tests/external-signed-ca-with-automatic-copy/install-server-with-external-ca-with-automatic-copy.yml:31]: The use of `ansible_fqdn` is also not correct here.

File: tests/external-signed-ca-with-manual-copy/install-server-with-external-ca-with-manual-copy.yml [hunk 4]
```diff
@@ -10,15 +10,14 @@
     state: present
 
   post_tasks:
-  - name: Copy CSR /root/ipa.csr from node to "{{ groups.ipaserver[0] + '-ipa.csr' }}"
+  - name: Copy CSR /root/ipa.csr from node to "{{ ansible_fqdn + '-ipa.csr' }}"
```
Comment (t-woerner) [inline tests/external-signed-ca-with-manual-copy/install-server-with-external-ca-with-manual-copy.yml:13]: Also here

File: tests/external-signed-ca-with-manual-copy/install-server-with-external-ca-with-manual-copy.yml [hunk 5]
```diff
@@ -10,15 +10,14 @@
...
     fetch:
       src: /root/ipa.csr
-      dest: "{{ groups.ipaserver[0] + '-ipa.csr' }}"
+      dest: "{{ ansible_fqdn + '-ipa.csr' }}"
```
Comment (t-woerner) [inline tests/external-signed-ca-with-manual-copy/install-server-with-external-ca-with-manual-copy.yml:16]: Also here

File: tests/external-signed-ca-with-manual-copy/install-server-with-external-ca-with-manual-copy.yml [hunk 6]
```diff
@@ -38,9 +37,9 @@
     #ipaserver_external_ca_file: "cacert.asc"
 
   pre_tasks:
-  - name: Copy "{{ groups.ipaserver[0] + '-chain.crt' }}" to /root/chain.crt on node
+  - name: Copy "{{ ansible_fqdn + '-chain.crt' }}" to /root/chain.crt on node
```
Comment (t-woerner) [inline tests/external-signed-ca-with-manual-copy/install-server-with-external-ca-with-manual-copy.yml:40]: Also here

File: tests/external-signed-ca-with-manual-copy/install-server-with-external-ca-with-manual-copy.yml [hunk 7]
```diff
@@ -38,9 +37,9 @@
...
     copy:
-      src: "{{ groups.ipaserver[0] + '-chain.crt' }}"
+      src: "{{ ansible_fqdn + '-chain.crt' }}"
```
Comment (t-woerner) [inline tests/external-signed-ca-with-manual-copy/install-server-with-external-ca-with-manual-copy.yml:42]: Also here

File: tests/dnsrecord/env_setup.yml [hunk 8]
```diff
@@ -18,8 +18,9 @@
     - "{{ zone_prefix_reverse_24 }}"
     - "{{ zone_prefix_reverse_16 }}"
     - "{{ zone_prefix_reverse_8 }}"
+    - "{{ zone_ipv6_reverse_workaround }}"
     - "{{ testzone }}"
-    - ip6.arpa.
+    - "{{ zone_ipv6_reverse }}"
```
Comment (t-woerner) [inline tests/dnsrecord/env_setup.yml:23]: Is this a bug fix?

File: tests/dnsrecord/env_cleanup.yml [hunk 9]
```diff
@@ -129,7 +129,7 @@
        - "{{ zone_prefix_reverse_24 }}"
        - "{{ zone_prefix_reverse_16 }}"
        - "{{ zone_prefix_reverse_8 }}"
+       - "{{ zone_ipv6_reverse }}"
+       - "{{ zone_ipv6_reverse_workaround }}"
```
Comment (t-woerner) [inline tests/dnsrecord/env_cleanup.yml:133]: Is this a bug fix?

File: tests/host/test_host_reverse.yml [hunk 10]
```diff
@@ -26,15 +25,29 @@
...
-                        join('.') }}"
+
+  - name: Set zone prefixes.
+    set_fact:
+      zone_ipv6_reverse: "ip6.arpa."
+      zone_ipv6_reverse_workaround: "d.f.ip6.arpa."
+      zone_prefix_reverse: "in-addr.arpa"
+      zone_prefix_reverse_8: "{{ ipv4_prefix.split('.')[2::-1] | join ('.') }}.in-addr.arpa"
+      zone_prefix_reverse_16: "{{ ipv4_prefix.split('.')[1::-1] | join ('.') }}.in-addr.arpa"
+      zone_prefix_reverse_24: "{{ ipv4_prefix.split('.')[::-1] | join ('.') }}.in-addr.arpa"
```
Comment (t-woerner) [inline tests/host/test_host_reverse.yml:36]: These are bigger changes that are not related to `_domain` or `_realm` and should be in a separate commit.

File: tests/external-signed-ca-with-automatic-copy/install-server-with-external-ca-with-automatic-copy.yml (hunk 2 above)
Comment (seocam) [reply tests/external-signed-ca-with-automatic-copy/install-server-with-external-ca-with-automatic-copy.yml:30]: Does this run on localhost? 3 lines above this change it says  `ipaserver`.

File: tests/external-signed-ca-with-manual-copy/install-server-with-external-ca-with-manual-copy.yml (hunk 4 above)
Comment (seocam) [reply tests/external-signed-ca-with-manual-copy/install-server-with-external-ca-with-manual-copy.yml:13]: Are you sure? We have `hosts: ipaserver` in this playbook.

File: tests/external-signed-ca-with-manual-copy/install-server-with-external-ca-with-manual-copy.yml (hunk 6 above)
Comment (seocam) [reply tests/external-signed-ca-with-manual-copy/install-server-with-external-ca-with-manual-copy.yml:40]: Same here.

File: tests/host/test_host_reverse.yml (hunk 10 above)
Comment (seocam) [reply tests/host/test_host_reverse.yml:36]: Ok. I fixed 4 bugs in a single commit. I'll create 1 commit for each fix.

**Comment by seocam** [comment]
@t-woerner I split the test fixes in 4 different commits (previously it was just one).

Can you please recheck your comments about the code that's running in `localhost`? In my understanding this code is running on a remote managed host named `ipaserver` but if you are sure that it's running on `localhost` I'll replace as you suggested.

Thanks!

File: tests/external-signed-ca-with-automatic-copy/install-server-with-external-ca-with-automatic-copy.yml (hunk 2 above)
Comment (t-woerner) [reply tests/external-signed-ca-with-automatic-copy/install-server-with-external-ca-with-automatic-copy.yml:30]: O.k. you are right that here we have `hosts: ipaserver` here. But as you said that `groups.ipaserver[0]` in Azure will not be FQDN. That means that `groups.ipaserver[0]` used to create the certificate and `ansible_fqdn` here are not the same. It would be good to use the same variable here and also to generate the certificate.

**Comment by seocam** [comment]
The goal of this PR is to make tests run in Azure without breaking any existing tests.

@t-woerner as we have discussed today I've removed the `ansible_fqdn` from the files:
* `tests/ca-less/install_server_without_ca.yml`
* `tests/external-signed-ca-with-manual-copy/install-server-with-external-ca-with-manual-copy.yml`
* `tests/external-signed-ca-with-automatic-copy/install-server-with-external-ca-with-automatic-copy.yml`

In order to improve that we can create a fact `ipaserver_fqdn` (or something similar) and use that everywhere instead of having differences in the test files. Since this goes beyond the scope of this PR I believe we can do that in a different PR.

File: tests/azure/templates/playbook_tests.yml [hunk 11]
```diff
@@ -0,0 +1,61 @@
...
+        "molecule[docker]>=3" \
+        "ansible${{ parameters.ansible_version }}" \
+        pytest \
+        pytest-split-tests
+    displayName: Install dependencies
+
+  - script: |
+      mkdir -p ~/.ansible/roles ~/.ansible/library ~/.ansible/module_utils
+      cp -a roles/* ~/.ansible/roles
+      cp -a plugins/modules/* ~/.ansible/library
```
Comment (t-woerner) [inline tests/azure/templates/playbook_tests.yml:39]: You can also directly copy the plugins folder, that will work also. ~/.ansible/plugins

**Comment by seocam** [comment]
Thanks @t-woerner!
I updated the azure pipelines to run with the new location .yml location and the first test run is already in progress.
